test/e2e/node/pod_resize.go

```diff
@@ -0,0 +1,1421 @@
+		}
+		if len(restartContainersExpected) == 0 {
+			return nil
+		}
+		for start := time.Now(); time.Since(start) < PollTimeout; time.Sleep(PollInterval) {
+			pod, err := podClient.Get(context.TODO(), pod.Name, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			restartedContainersCount := 0
+			for _, cName := range restartContainersExpected {
+				cs, _ := podutil.GetContainerStatus(pod.Status.ContainerStatuses, cName)
+				if cs.RestartCount < 1 {
+					break
+				}
+				restartedContainersCount++
+			}
+			if restartedContainersCount == len(restartContainersExpected) {
+				return nil
+			}
+		}
+		return fmt.Errorf("timed out waiting for expected container restart")
+	}
+	waitPodAllocationsEqualsExpected := func() (*v1.Pod, error) {
+		for start := time.Now(); time.Since(start) < PollTimeout; time.Sleep(PollInterval) {
+			pod, err := podClient.Get(context.TODO(), pod.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+			if verifyPodAllocations(pod, expectedContainers, false) == false {
+				continue
+			}
+			return pod, nil
+		}
+		return nil, fmt.Errorf("timed out waiting for pod resource allocation values to match expected")
+	}
+	waitContainerCgroupValuesEqualsExpected := func() error {
+		for start := time.Now(); time.Since(start) < PollTimeout; time.Sleep(PollInterval) {
+			if verifyPodContainersCgroupValues(patchedPod, expectedContainers, false) == false {
+				continue
+			}
+			return nil
+		}
+		return fmt.Errorf("timed out waiting for container cgroup values to match expected")
+	}
+	waitPodStatusResourcesEqualSpecResources := func() (*v1.Pod, error) {
+		for start := time.Now(); time.Since(start) < PollTimeout; time.Sleep(PollInterval) {
+			pod, err := podClient.Get(context.TODO(), pod.Name, metav1.GetOptions{})
+			if err != nil {
+				return nil, err
+			}
+			differs := false
+			for idx, c := range pod.Spec.Containers {
+				if diff.ObjectDiff(c.Resources, *pod.Status.ContainerStatuses[idx].Resources) != "" {
+					differs = true
+					break
+				}
+			}
+			if differs {
+				continue
+			}
+			return pod, nil
+		}
+		return nil, fmt.Errorf("timed out waiting for pod spec resources to match pod status resources")
+	}
+	rsErr := waitForContainerRestart()
+	framework.ExpectNoError(rsErr, "failed waiting for expected container restart")
+	// Wait for pod resource allocations to equal expected values after resize
+	resizedPod, aErr := waitPodAllocationsEqualsExpected()
+	framework.ExpectNoError(aErr, "failed to verify pod resource allocation values equals expected values")
+	// Wait for container cgroup values to equal expected cgroup values after resize
+	cErr := waitContainerCgroupValuesEqualsExpected()
+	framework.ExpectNoError(cErr, "failed to verify container cgroup values equals expected values")
+	//TODO(vinaykul,InPlacePodVerticalScaling): Remove featureGatePostAlpha upon exiting Alpha.
+	//                containerd needs to add CRI support before Beta (See Node KEP #2273)
```

Comment:
Is there support in cri-o already?

Rather than gating validation on alpha, we might want to gate based on whether the feature is enabled?

If we do that, plus the move to test/e2e_node/..., we can fairly easily introduce a test lane when experimental support lands, without needing to make code changes.

Reply:
HI @endocrimes Hope you're feeling much better now!

I don't believe we have CRI-O yet (I've not had the time to jump into their meetings and discuss my needs with them) But @ruiwen-zhao  has a [draft PR](https://github.com/containerd/containerd/pull/6517) ready for containerd, which is our immediate priority since we use containerd in CI. (I had added support in dockershim but that got deprecated)

And this check is a safeguard against entering beta without containerd support "in-play" and a [tracked TODO](https://github.com/vinaykul/kubernetes/wiki/In-Place-Pod-Vertical-Scaling-Issues-and-Status) to remove it after @ruiwen-zhao PR lands and is picked up by us. This test does run today as part of e2e-gce-alpha-features job and verifies successful resize by looking at the cgroup values for each test. Once the containerd change is in, we will enable this. If it is enabled without the containerd change, the test will fail.

Reply:
In retrospect, I should have pushed for CRI changes go in a preceding release to allow for CRI changes to happen (which is why I had a separate KEP for it). But at some point, my (flawed) thinking was it would become a third wheel without the core feature - and much of the core feature work was in a bit of flux....  And it wouldn't have been that big of an issue if I had managed to land it <1.23 (before dockershim removal) as these tests would have just worked E2E.

But I'm tracking @ruiwen-zhao PR closely and we probably could use a bit of 'carnival barking' at the next KubeCon to get as many runtimes as possible to support this. 😃 As soon as containerd support is in, this check is gone, and hopefully many more runtimes will support it by GA (1.28)

Reply:
the CRI-O team is prepared to support it once the CRI changes are merged :slightly_smiling_face: